### PR TITLE
Campaign update UI

### LINKED ISSUE #7683: a8n/web: Allow updating the CampaignPlan and other fields of a Campaign
After https://github.com/sourcegraph/sourcegraph/pull/7328 we need to implement the UI that allows a user to update a Campaign with a new CampaignPlan. See the discussion in #7328 for more details.

- [ ] Add a section to the docs at https://docs.sourcegraph.com/user/automation about updating a campaign. Mention all of the kinds of operations that can occur due to an update (start from the bullet list under Proposal in [RFC 95](https://docs.google.com/document/d/1_0fOwbmIov3utmpFa2cs_3qSM3HGO3mNLTUVB8sZxIg/edit#heading=h.umigb8il1vw3) and ask @mrnugget to make sure any other edge cases he considered are mentioned if users would care).

### PR DESCRIPTION
**Pull request by eseliger**
Part of #7683

### CODE REVIEW

**Comment by eseliger** [comment]
image

this is the result @mrnugget and I came up with

**Comment by eseliger** [comment]
Update: I addressed the update workflow UI in this PR, see gif 
image

Related: https://github.com/sourcegraph/src-cli/pull/122

File: web/src/enterprise/campaigns/detail/CampaignUpdateSelection.tsx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+interface Props {
+    onSelect: (campaign: Pick<GQL.ICampaign, 'id'>) => void
+    history: H.History
+    location: H.Location
+    className?: string
+}
+
+/**
+ * A list of a campaign's changesets changed over a new plan
```
Comment (mrnugget) [inline web/src/enterprise/campaigns/detail/CampaignUpdateSelection.tsx]: That comment sounds wrong?

File: web/src/enterprise/campaigns/detail/CampaignUpdateSelection.tsx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+                GQL.ICampaign,
+                'id' | 'closedAt' | 'name' | 'description' | 'changesets' | 'changesetPlans' | 'createdAt'
+            >,
+            Omit<CampaignNodeProps, 'node'>
+        >
+            history={history}
+            location={location}
+            nodeComponent={CampaignNode}
+            nodeComponentProps={{
+                selection: { buttonLabel: 'Select', enabled: true, onSelect },
```
Comment (mrnugget) [inline web/src/enterprise/campaigns/detail/CampaignUpdateSelection.tsx]:
```suggestion
                selection: { buttonLabel: 'Update this campaign...', enabled: true, onSelect },
```

File: web/src/enterprise/campaigns/detail/CampaignUpdateSelection.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    className?: string
+}
+
+/**
+ * A list of a campaign's changesets changed over a new plan
+ */
+export const CampaignUpdateSelection: React.FunctionComponent<Props> = ({ history, location, onSelect }) => (
+    <>
+        <PageTitle title="Update campaign" />
+        <h1>Update campaign</h1>
```
Comment (mrnugget) [inline web/src/enterprise/campaigns/detail/CampaignUpdateSelection.tsx]:
```suggestion
        <h1>Select campaign to update</h1>
```

File: web/src/enterprise/campaigns/detail/CampaignUpdateSelection.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        <p>Select the campaign to update from the list of campaigns below:</p>
```
Comment (mrnugget) [inline web/src/enterprise/campaigns/detail/CampaignUpdateSelection.tsx]:
```suggestion
        <p>Select a campaign to update from the list below to preview which changes will be made on the code hosts:</p>
```

File: web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx [hunk 5]
```diff
@@ -58,7 +58,9 @@ export const ChangesetNode: React.FunctionComponent<ChangesetNodeProps> = ({
             setPublishError(undefined)
             setIsLoading(true)
             await _publishChangeset(node.id)
-            campaignUpdates.next()
+            if (campaignUpdates) {
```
Comment (lguychard) [inline web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx]: What is the logic behind not always passing in the `campaignUpdates` Subject?

File: web/src/enterprise/campaigns/detail/CampaignStatus.tsx [hunk 6]
```diff
@@ -73,26 +73,27 @@ export const CampaignStatus: React.FunctionComponent<Props> = ({ campaign, statu
...
-                        {campaign.__typename === 'Campaign' ? 'Creation' : 'Preview'} {status.state.toLocaleLowerCase()}
+                        <CheckCircleIcon className="icon-inline text-success mr-1 e2e-preview-success" /> Creation
+                        completed
                     </div>
                 )
             )}
+            {status.state === GQL.BackgroundProcessState.ERRORED && (
+                <>
+                    <AlertCircleIcon className="icon-inline text-danger mr-1" />
+                    Creation errored
```
Comment (lguychard) [inline web/src/enterprise/campaigns/detail/CampaignStatus.tsx]:
```suggestion
                    Error creating campaign
```

**Comment by mrnugget** [comment]
Just making sure that this is not lost: the ticket https://github.com/sourcegraph/sourcegraph/issues/7683 also includes updating the user documentation.

File: web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx (hunk 5 above)
Comment (eseliger) [reply web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx]: the update UI doesn't have a update mechanism as the plan never changes 🤔

Comment (lguychard) [reply web/src/enterprise/campaigns/detail/changesets/ChangesetNode.tsx]: Makes sense. Thanks!

**Pull request merged**